Add getting started page to user manual

pledbrook opened the pull request:
This also introduces a page that covers some important, core Gradle concepts
that all users should be aware of. And a primer to help understand how the Gradle API maps to Groovy build script syntax.

eskatos commented on subprojects/docs/src/docs/userguide/five_things.adoc:
```diff
@@ -0,0 +1,125 @@
++
+When you want the build to do some work that an existing task can't do, you can simply write your own task type. It's typically best to put the source file for a custom task type in the <<organizing_gradle_projects#sec:build_sources,_buildSrc_>> directory or in a packaged plugin. Then you can use the custom task type just like any of the Gradle-provided ones.
+
+ 4. Custom conventions.
```
:x: We should not encourage the use of "conventions" to extend the DSL but of "extensions" instead.
See https://guides.gradle.org/implementing-gradle-plugins/#extensions_vs_conventions

pledbrook replied:
I'm talking about Gradle's internal conventions model, but generic conventions, such as directory structures and names. Since this chapter is primarily targeted at newcomers to Gradle, I'm pretty sure they will interpret "convention" in the traditional sense.

eskatos replied:
Riiight, I mixed my mind :)

eskatos commented on subprojects/docs/src/docs/userguide/five_things.adoc:
```diff
@@ -0,0 +1,125 @@
+
+Gradle provides several mechanisms that allow you to extend it, from the very simple to the very powerful:
+
+ 1. Extra properties on <<writing_build_scripts#sec:extra_properties,projects>> and <<writing_build_scripts#sec:extra_task_properties,tasks>>.
```
(extra) properties are handy but they are untyped and it's hard to know where they come from. Most often there're better ways to do what they are used for, with more modeling.

I would not put them first on that enumeration.

pledbrook replied:
The order is from simplest to most complex/work involved. I'm not that keen on changing the order, but as @eriwen suggests, perhaps your concerns can be addressed a little with unnumbered bullets. I can see the numbering implying some sense of priority over and above the item being first in the list.

eskatos replied:
Sounds good 👍

pioterj replied:
Wouldn't it make more sense to order by how common they are and start with custom tasks?

pledbrook replied:
Maybe. I don't have any insight into what's most common though. It's probably worth ordering them in such a way as our preferred options come first. I could be persuaded either way.

pledbrook replied:
I changed the order to something closer to most common -> less common.

eskatos commented on subprojects/docs/src/docs/userguide/groovy_build_script_primer.adoc:
```diff
@@ -0,0 +1,224 @@
+// limitations under the License.
+
+[[groovy_build_script_primer]]
+= A Groovy Build Script Primer
```
It would be nice to cover how to interact with containers.
`tasks`, `configurations`, `sourceSets` etc.. and the ones contributed by plugins are based on the same abstraction and the Groovy DSL provides sugar to interact with them.

pledbrook replied:
What sorts of things are you thinking about beyond the properties and methods documented in the API docs? Happy to add anything that helps interpret what's happening in a build script.

eskatos replied:
Right, there are reasonable API docs for the containers. I think that given these containers of named domain objects are very frequent in the DSL and that community plugins also make use of these containers to build their DSL we could consider explaining their usage in a primer.

pledbrook replied:
Do you have specific examples of what to cover? I recognise that it's important to cover domain object containers somewhere, but is there anything specific to the Groovy DSL that is worth tackling here?

pledbrook replied:
@eskatos Wondering if you have any feedback on ☝️

eskatos replied:
@pledbrook, sorry for not following up earlier

The Groovy DSL adds some magic on top of the container API.

```groovy
def container = container(Something)
container.elementName // access element by name
container.configure {
    foo { // maybeCreate element by name and configure it
        gizmo = "else"
    }
}
```

and for polymorphic containers too.

The Kotlin DSL provides different conveniences as shown for the `tasks` container in the [migration guide](https://guides.gradle.org/migrating-build-logic-from-groovy-to-kotlin/#configuring-tasks).

But maybe this should go into a place dedicated to interacting with containers instead of here, showing both Groovy and Kotlin samples.

pledbrook replied:
No problem. Although there is some magic here, it takes the form of normal syntax: property access and code blocks/closures. It may be worth including some general info on dynamic properties and methods, but keep the specifics with the documentation for the relevant features. Of course, we'd need to add something on domain object containers because I don't think there is anything right now.

pledbrook replied:
I've raised [an issue](https://github.com/gradle/dotorg-docs/issues/288) about this.

pledbrook commented on subprojects/docs/src/docs/userguide/five_things.adoc:
```diff
@@ -0,0 +1,125 @@
+Tasks themselves consist of:
+
+ * Actions — pieces of work that do something, like copy files or compile source
+ * Inputs — values, files and directories that the actions use or operate on
```
This was suggested by @oehme in another issue, so I'd like to get his feedback.

From my perspective, task dependencies are already covered by the discussion about the task graph. This list is more about the work each individual task does in isolation.

oehme replied:
I think this is good as-is

pledbrook commented on subprojects/docs/src/docs/userguide/five_things.adoc:
```diff
@@ -0,0 +1,125 @@
+
+In fact, all of the above are optional depending on what the task needs to do. Some tasks — such as the <<base_plugin#sec:base_tasks,standard lifecycle tasks>> — don't even have any actions. They simply aggregate multiple tasks together as a convenience.
+
+NOTE: You choose which task to run. Save time by specifying the task that does what you need, but no more than that. If you just want to run the unit tests, choose the task that does that — typically `test`. If you want to package an application, most builds have an `assemble` task for that.
```
Why the wording change? If you think the current wording is harder to parse, fair enough. But the current wording is more of a directive, which is how I interpreted @oehme's intent on this one.

I don't understand your comment about tasks being abstracted from the user view. What do you mean by that? I'm all for being friendlier and more helpful to IDE users, I just don't understand how this note is affected.

eriwen replied:
This is good as-is.

I'm trying (badly) to point out that many users just run Gradle through the IDE in such a way that they don't actually know or care what Gradle tasks they're invoking.

I am mostly asking that we acknowledge that in the text, but not that we should change this NOTE about it.

pledbrook replied:
> I'm trying (badly) to point out that many users just run Gradle through the IDE in such a way that they don't actually know or care what Gradle tasks they're invoking.

@eriwen Is that true? How does the build get run then? I'm afraid I'm still not familiar enough with IDE usage.

eriwen replied:
IDEs use the Tooling API to invoke Gradle (though in certain circumstances they invoke underlying tools like `javac` instead for performance). 

You can see this in action when clicking the green arrow to the left of a test name, for example.

pledbrook replied:
Right, but whether it's via the Tooling API or the command line, the IDE is still running tasks, right? Is the integration with things like Project > Build, Project > Compile and unit tests sufficient for the user _not_ to interact with tasks directly at some point? I guess that depends on the IDE and what most developers need to do.

pledbrook commented on subprojects/docs/src/docs/userguide/five_things.adoc:
```diff
@@ -0,0 +1,125 @@
+
+NOTE: You choose which task to run. Save time by specifying the task that does what you need, but no more than that. If you just want to run the unit tests, choose the task that does that — typically `test`. If you want to package an application, most builds have an `assemble` task for that.
+
+One way to customize a build is to add extra steps to it. This is easy to do in Gradle: define your own tasks and insert them into the task graph via the <<tutorial_using_tasks#sec:task_dependencies,task dependency mechanism>>. You can have your tasks depend on any other task or configure existing tasks to depend on yours. The main restriction is that you cannot create any circular dependencies w ...
```
I'm a little hesitant to even include this paragraph. This is not a tutorial on how to write tasks and I want to keep it as focused as possible. It's just that I don't think it's obvious to people how to integrate their tasks into an existing task graph.

It's really a question of how important the information is to understanding how Gradle works. Talking about `doLast()` could end up adding an extra paragraph or two with little value to someone who knows little about Gradle at this point. Happy to be persuaded otherwise though.

eriwen replied:
You may be able to remove this paragraph without losing anything critical, because I think it's implied and covered in docs linked to.

pledbrook replied:
@eriwen I've removed the paragraph, but added a new, similar statement to an earlier paragraph (the one starting "Almost any build process can be modeled ..."). I'm not sure about the exact wording, but I think it usefully highlights how a task graph comes into being.

pledbrook commented on subprojects/docs/src/docs/userguide/five_things.adoc:
```diff
@@ -0,0 +1,125 @@
+
+One last thing: Gradle's <<more_about_tasks#sec:up_to_date_checks,incremental build>> support is robust and reliable, so keep your builds running fast by avoiding the `clean` task unless you actually do want to perform a clean. 
+
+== 3. Gradle has distinct build phases
```
I don't know. "Distinct" just implies "separate" to me. Or clearly delineated. I also think it implies that they have distinct roles, which "multiple" doesn't. Could be persuaded otherwise, but "multiple" doesn't work as well for me. And in large part, the phases are sequential in a basic understanding of Gradle, which is as far as we want to take people with this chapter.

eriwen replied:
Okay. I don't feel too strongly about this.

pledbrook replied:
@eriwen I've tried with "fixed" instead of "distinct". It's a completely different meaning, but may be more useful information to users. What do you think?

pledbrook commented on subprojects/docs/src/docs/userguide/five_things.adoc:
```diff
@@ -0,0 +1,125 @@
+
+NOTE: Gradle's build phases are not like Maven's phases. Maven uses its phases to divide the build execution into multiple stages. They serve a similar role to Gradle's task graph, although less flexibly.
+
+When you are authoring or modifying a build, it's important to recognize which bits are evaluated during which phases. Otherwise you will end up with a build that doesn't work as expected. For example, if you set a property within the execution phase, any part of the build that reads that property during the configuration phase won't see the new value.
```
If these types of problem are covered in Troubleshooting, absolutely.

eriwen replied:
On second thought, maybe this subsection is better: https://docs.gradle.org/current/userguide/authoring_maintainable_build_scripts.html#sec:minimize_logic_executed_configuration_phase

pledbrook replied:
I removed this content as it was delving too much into the weeds for this type of chapter. However, I replaced it with a paragraph that does indeed link to that section.

pledbrook commented on subprojects/docs/src/docs/userguide/five_things.adoc:
```diff
@@ -0,0 +1,125 @@
+    include "**/*.txt"
+}
+----
+<1> Evaluated during _execution_ phase
```
What's an "execution" action? Or do you mean `doFirst()` simply registers a task action?

pledbrook replied:
It doesn't help that we use `Action` for that interface which has nothing to do with task actions :smile:

pledbrook replied:
I removed this example.

pledbrook commented on subprojects/docs/src/docs/userguide/five_things.adoc:
```diff
@@ -0,0 +1,125 @@
+<1> Evaluated during _execution_ phase
+<2> Evaluated during _configuration_ phase
+
+This used to be more of a problem when the `<<` syntax for defining a task was more common, as it implicitly declared a task action — something that is evaluated/run during the execution phase. Users would often attempt to configure the task within that definition and wonder why it wasn't working.
```
Good point. Either that or remove the information. Probably best to keep it as there are probably old builds around that still use `<<`.

pledbrook replied:
In the end I removed the content completely (see previous thread).

pledbrook commented on subprojects/docs/src/docs/userguide/five_things.adoc:
```diff
@@ -0,0 +1,125 @@
+
+Gradle provides several mechanisms that allow you to extend it, from the very simple to the very powerful:
+
+ 1. Extra properties on <<writing_build_scripts#sec:extra_properties,projects>> and <<writing_build_scripts#sec:extra_task_properties,tasks>>.
```
See comment in reply to Paul. I think unnumbered might be more appropriate here.

pledbrook commented on subprojects/docs/src/docs/userguide/five_things.adoc:
```diff
@@ -0,0 +1,125 @@
+
+NOTE: As Gradle runs on the JVM, build scripts can also use the standard {javaApi}[Java API]. Groovy build scripts can additionally use the Groovy APIs, while Kotlin build scripts can use the Kotlin ones.
+
+++[++This section needs more detail, or a dedicated chapter that it can link to. I don't think anything exists right now.]
```
Sorry, I missed pushing my latest changes with the link to the Groovy Build Script Primer. That's where I want the information on syntax technicalities. 

The comment about best practices would be more suitable for the section on extending Gradle I think.

pledbrook commented on subprojects/docs/src/docs/userguide/getting-started.adoc:
```diff
@@ -1,6 +0,0 @@
-include::installation.adoc[]
-
-[[sec:installation_getting_help]]
-=== Getting help
```
OK. Can we keep them separate? I'm assuming that the role of the two documents is different. I could reinstate _getting-started.adoc_, although it would be confusing having both under _src/docs/userguide_. I don't really understand why a file that goes into the distribution is under _src/docs/userguide_ unless it's the user manual itself.

pledbrook commented on subprojects/docs/src/docs/userguide/getting_started.adoc:
```diff
@@ -0,0 +1,67 @@
+
+For more specialized builds or those with significant customizations, you should ideally have access to documentation on how to run the build and what build properties you can configure.
+
+== How do I migrate from another build tool
```
Probably

pledbrook commented on subprojects/docs/src/docs/userguide/getting_started.adoc:
```diff
@@ -0,0 +1,67 @@
+
+ * <<five_things#five_things,5 things you need to know about Gradle>>
+
+== Short tutorials
```
Good point. Not sure yet. I don't think "Short tutorials" can usefully be replaced by a question. I'll probably end up rewording some of the other titles so that they aren't questions. Perhaps I'll just leave "What is Gradle?" as a standalone question. I think that would look fine.

pledbrook commented on subprojects/docs/src/docs/userguide/getting_started.adoc:
```diff
@@ -0,0 +1,67 @@
+ * {guidesUrl}/creating-new-gradle-builds/[Creating a basic Gradle build]
+ * {guidesUrl}/building-android-apps/[Building Android apps]
+ * {guidesUrl}/building-java-libraries/[Building Java libraries]
+ * {guidesUrl}/building-cpp-libraries/[Building C++ libraries]
```
/me shrugs. I was just trying to think of the main ecosystems. If we add Kotlin, then the obvious question is why not Scala, Groovy, Swift, assembler (and so on) as well? I don't feel strongly about this, other than I don't want it to turn into a long list.

eriwen replied:
@pledbrook Yes, please. Specifically Kotlin because the vast majority of Kotlin developers use Gradle. This is different from other languages list. We should consider removing C++ here based on that criteria, because that guide uses the deprecated software model.

pledbrook replied:
I'd argue that's a different criterion. Regardless, does the guide need updating for the new C++ plugins? Or are those plugins not ready for prime time yet?

pledbrook commented on subprojects/docs/src/docs/userguide/getting_started.adoc:
```diff
@@ -0,0 +1,67 @@
+== How do I migrate from another build tool
+
+ * {guidesUrl}/migrating-from-maven/[Migrating from Apache Maven]
+
```
I think I intended to have a section geared more towards build authors, but this had been gestating so long I wanted to make it available.

That said, I don't think we have a great learning path for build authors right now beyond the various Getting Started Guides. So I'm not sure what would go into such a section.

I think it's reasonable to link to third-party tool integration and extending Gradle, as those are topics reader may want to "get started" with.

pledbrook replied:
I've now added a section for build authors. Not sure whether to add something for integrating Gradle.

pledbrook commented on subprojects/docs/src/docs/userguide/groovy_build_script_primer.adoc:
```diff
@@ -0,0 +1,224 @@
+[[groovy:properties]]
+== Properties
+
+[source,groovy]
```
Not unless there are blank lines in the source code. Probably makes sense to include them regardless, just for consistency and to avoid confusion.

pledbrook commented on subprojects/docs/src/docs/userguide/groovy_build_script_primer.adoc:
```diff
@@ -0,0 +1,224 @@
+[source,groovy]
+<obj>.<name>                // Get a property value
+<obj>.<name> = <value>      // Set a property to a new value
+"$<name>"                   // Embed a property value in a string
```
Yeah, not sure. I'm used to angle brackets in other docs demarcating abstractions. I'm expecting the examples that follow to make it clear the angle brackets aren't part of the syntax.

The comments could probably be improved to help and I'm wondering whether to use annotations. I like the inline comment approach when the comment is short, though.

eriwen replied:
I don't feel strongly about the angle brackets, so we can keep them. +1 on the inline comments

pledbrook commented on subprojects/docs/src/docs/userguide/getting_started.adoc:
```diff
@@ -0,0 +1,67 @@
+
+Everyone has to start somewhere and if you're new to Gradle, then this is a good place to begin.
+
+== What is Gradle?
```
My current thinking is to make User Manual Home more of a docs home that explains what each section of the docs is for and help folks navigate around them.

We also need to consider that http://gradle.org/ should have a "What is Gradle?" in some shape or form. This page could then link to that. Although I would expect that to more marketing-oriented, whereas this should be more technical with a slightly different focus.

eriwen replied:
Heck yeah gradle.org should host the authoritative answer to "What is Gradle?". I like how you think about the User Manual Home page. I'm not sure I can help much with such a page given the near-absurd amount of stuff we want to get done before 5.0.

BTW, we now have hosted search available from DocSearch that I'd really like to include in Gradle 5.0.

eriwen replied:
This still doesn't quite feel right. I think what Bazel does here in terms of organizing the information is pretty darn superior to how the User Manual Home + Getting Started is organized. 

We can ship this change before doing much with userguide.adoc, but I'd consider making changes to the home page a blocker for 5.0.

pioterj commented on subprojects/docs/src/docs/userguide/five_things.adoc:
```diff
@@ -0,0 +1,126 @@
+ 2. Configuration
+ 3. Execution
+
+In well-designed build scripts, the vast majority of what you see is evaluated during the configuration phase. Even so, some elements are likely to be evaluated during the execution phase, particularly the contents of `doLast {}` and `doFirst {}` blocks.
```
This paragraph is problematic. Rather than explaining what each phase is it already talks about their consequences and assumes prior understanding. It may also be read as the majority of work in well-designed build happens in configuration phase which is of course not the case. I'd just replace this paragraph with a link to [build lifecycle documentation](https://docs.gradle.org/current/userguide/build_lifecycle.html).

pledbrook replied:
The aim is to highlight how an understanding of the build lifecycle can help with authoring well-behaved and maintainable build scripts. I don't think linking to that chapter is sufficient to explain the why of this item. I agree this needs a rewrite though.

pioterj commented on subprojects/docs/src/docs/userguide/five_things.adoc:
```diff
@@ -0,0 +1,126 @@
+// limitations under the License.
+
+[[five_things]]
+= Five things you need to know about Gradle
```
How about we make it six and add build scans? :) cc @eriwen

pledbrook replied:
I think it's better to incorporate build scans where there is obvious value. They aren't necessary to use Gradle properly and inserting them here risks making it look like marketing. That said, I think we can add them to the main Getting Started page.

pledbrook replied:
I added build scans as an item under What is Gradle? in the Getting Started page.

pledbrook commented on subprojects/docs/src/docs/userguide/five_things.adoc:
```diff
@@ -0,0 +1,126 @@
+
+When you are authoring or modifying a build, it's important to recognize which bits are evaluated during which phases. Otherwise you will end up with a build that doesn't work as expected. For example, if you set a property within the execution phase, any part of the build that reads that property during the configuration phase won't see the new value.
+
+Consider the following (contrived) example that attempts to initialise a `Copy` task from a properties file. You might try this if the properties file is generated by another task. The problem is that `into` must be set during the configuration phase, otherwise the task won't run. Setting the value within a `doFirst {}` block fails as it happens too late.
```
Agreed, this does get too much into the weeds. As stated above, though, I want to keep some explanation of why it's important to understand the build lifecycle, and hence give users a reason to read the Build Lifecycle chapter.

oehme commented on subprojects/docs/src/docs/userguide/five_things.adoc:
```diff
@@ -0,0 +1,126 @@
+
+This doesn't mean you have to do a lot of work to create a build. Gradle makes it easy to build common types of project — say Java libraries — by adding a layer of conventions and prebuilt functionality through <<plugins#plugins,_plugins_>>. You can even create and publish custom plugins to encapsulate your own conventions and build functionality.
+
+== 2. The core model is based on a network of tasks
```
task graph instead of "network of tasks"?

pledbrook replied:
I'm not sure that "task graph" is meaningful to anyone that's new to Gradle. However, I agree that "network of tasks" isn't the best name either. I'm going to try "based on tasks" instead.

oehme commented on subprojects/docs/src/docs/userguide/five_things.adoc:
```diff
@@ -0,0 +1,126 @@
+
+ * Actions — pieces of work that do something, like copy files or compile source
+ * Inputs — values, files and directories that the actions use or operate on
+ * Outputs — files and directories that the actions modify or generate
```
modifies or generates
or
will modify or generate

oehme replied:
Sorry I misread it as "action" (singular).

oehme commented on subprojects/docs/src/docs/userguide/five_things.adoc:
```diff
@@ -0,0 +1,126 @@
+ 2. Configuration
+ 3. Execution
+
+In well-designed build scripts, the vast majority of what you see is evaluated during the configuration phase. Even so, some elements are likely to be evaluated during the execution phase, particularly the contents of `doLast {}` and `doFirst {}` blocks.
```
If find this phrase problematic, as it implies that "as much as possible should happen at configuration time", the exact opposite of what we want. I think I know what you meant, but it's not clear to a newcomer.

pledbrook replied:
Correct. Will push soon...

oehme replied:
This doesn't really apply given new developments like lazy task configuration. Many of the blocks you see in a build scripts don't get executed at all in any given build. I would put the whole thing on its head and say: 

In order to get from startup to the actual task execution as quickly as possible, building a model of your projects and tasks needs to be fast. It is therefor important to defer any expensive computation (like resolving dependencies and walking file trees) to task execution time.

oehme commented on subprojects/docs/src/docs/userguide/five_things.adoc:
```diff
@@ -0,0 +1,126 @@
+.Configuring a task within an action (execution phase) won't work
+----
+task copyFiles(type: Copy) {
+    doFirst {
```
❌ This will actually work just fine. It only becomes problematic when you try to make changes that would have affected the task graph. This is not such a change though.

oehme commented on subprojects/docs/src/docs/userguide/five_things.adoc:
```diff
@@ -0,0 +1,126 @@
+
+NOTE: Gradle's build phases are not like Maven's phases. Maven uses its phases to divide the build execution into multiple stages. They serve a similar role to Gradle's task graph, although less flexibly.
+
+When you are authoring or modifying a build, it's important to recognize which bits are evaluated during which phases. Otherwise you will end up with a build that doesn't work as expected. For example, if you set a property within the execution phase, any part of the build that reads that property during the configuration phase won't see the new value.
```
I think we should mention the opposite problem, which is much more common - Accidentally executing stuff during the configuration phase, e.g. by forgetting `doLast` or by forcing resolution of dependencies.

oehme commented on subprojects/docs/src/docs/userguide/five_things.adoc:
```diff
@@ -0,0 +1,126 @@
+
+ 5. {guidesUrl}/implementing-gradle-plugins/#modeling_dsl_like_apis[A custom model].
++
+Gradle allows you to introduce new concepts into a build beyond tasks, files and dependency configurations. You can see this with most language plugins, which add the concept of <<building_java_projects#sec:java_source_sets,_source sets_>> to a build. Appropriate modeling of a build process can greatly improve a build's ease of use and its efficiency. But new model elements should be used sparing ...
```
I don't agree with the last sentence. Most incomprehensible builds are due to a lock of modeling, not too much.

pledbrook replied:
So what I'm trying to get at is the classic problem of introducing too many abstractions via extensions or other means. I'm reminded of LinkedIn's Ant build which was cleanly done, but you basically had to learn a new build tool due to the abstractions it introduced.

Perhaps it's not worth adding the qualifier here, but rather focus on new abstractions that provide real value.

oehme replied:
The Android plugin is a great example of an abstraction that's basically a "new build tool" on top of Gradle. It introduces build types, flavors, variants, transformations etc. But at the same time it hides a lot of other things from users and makes it easier to integrate logic at a higher level. I think this is a good approach if your domain is complex enough to warrant the cost.

pledbrook replied:
I've removed the last sentence as it's veering into best practice, which isn't really appropriate for this page and would require substantial coverage. I think the "Appropriate modeling" is a sufficient qualifier here.

oehme commented on subprojects/docs/src/docs/userguide/getting_started.adoc:
```diff
@@ -0,0 +1,67 @@
+If you follow any of the tutorials linked above, you will execute a Gradle build. But what do you do if you're given a Gradle build without any instructions?
+
+ 1. Determine whether the project has a Gradle wrapper and <<gradle_wrapper#sec:using_wrapper,use it if it's there>>.
+ 2. Determine whether it's a single-project or multi-project build.
```
How is that important for the user?

pledbrook replied:
Because the user wants to run tasks, and the tasks are split between projects. It's also important to understand whether the build file in the root of the project is a normal one or geared towards configuring subprojects.

oehme replied:
If the user wants to know about the tasks, they can use `gradle tasks`. As for build files, I was under the impression that this section is aimed at the end user who just wants to run the build, not edit it.

pledbrook commented:
I've incorporated much of the feedback. There's still work to be done, but I may break that into separate pull requests.

@eriwen I've modified the navigation. There is now just a single _Getting Started_ link in the left nav. The links that were under the Getting Started heading are now in the page itself. I think the current approach makes sense, but I can also see adding something like the Maven migration guide direct to the left nav as well.

eriwen commented on subprojects/docs/src/docs/userguide/getting_started.adoc:
```diff
@@ -0,0 +1,69 @@
+
+For more specialized builds or those with significant customizations, you should ideally have access to documentation on how to run the build and what build properties you can configure.
+
+== Migrating from other build tools
```
❓ This subsection feels a little out of place at the end here with just one bullet point. 

Perhaps we could add a "to learn how to migrate your project to Gradle, follow the appropriate guide" kind of paragraph?

pledbrook replied:
If the link remains in the left nav — something I'm fine with — I'm tempted to remove the section if we can't provide any useful information on migrating from other build tools. Are there any "appropriate guides" we can link to?

pledbrook replied:
Removed. It's linked from my suggested home page though and should remain in the left navigation.

eriwen commented on subprojects/docs/src/main/resources/header.html:
```diff
@@ -110,13 +111,6 @@ <h3 id="reference">Reference</h3>
             <li id="third-party-integration"><a href="/userguide/third_party_integration.html">Gradle &amp; Third-party Tools</a></li>
         </ul>
 
-        <h3 id="getting-started">Getting Started</h3>
```
:x: I think that significantly fewer folks will see the "Creating build scans" and "Migrating from Maven" guides since they have been demoted a ways in the organizational hierarchy.

I'd like to share an idea of how we can compromise here, and I welcome your ideas, too.

pledbrook replied:
As mentioned in a previous comment, I'm fine with the migration guide being in the left nav. I wanted it alongside the upgrade link as you have below.

The "Creating build scans" link still seems like an orphan to me that doesn't really belong at such a high level of the navigation. I would prefer to move it to the _Running Gradle builds_ section and tackle it from a more user-friendly perspective, i.e. one that users can readily understand at first look without first having build scans explained to them. I remember suggesting something like "Gaining insights into your build" and I still think we would be better off with an approach along those lines, even if a different phrase is used.

eriwen commented on subprojects/docs/src/main/resources/header.html:
```diff
@@ -95,6 +95,7 @@
             </div></li>
             <li><a href="/release-notes.html">Release Notes</a></li>
             <li><a href="/userguide/installation.html">Installing Gradle</a></li>
+            <li><a href="/userguide/getting_started.html">Getting Started</a></li>
```
The top left nav now is getting quite cluttered. I think some of these links could go below and we can break this up a bit.

eriwen replied:
How about this?

![screenshot 2018-09-13 17 34 24](https://user-images.githubusercontent.com/51534/45522980-5afa1780-b77b-11e8-946e-4782cbeff9c2.png)

eriwen replied:
Note that I had to do a little bit of CSS trickery to make a gap. I'm happy to handle the nav separately if you like.

pledbrook replied:
@eriwen Looking at the proposal again, where would release notes go?

I'm OK with losing the "User Manual Home" link as long as a _Documentation_ link is retained in the top nav. I see that on the nightly docs, there is no such link at all, not even the old drop-down list.

eriwen commented on subprojects/docs/src/docs/userguide/five_things.adoc:
```diff
@@ -0,0 +1,112 @@
+
+Gradle allows you to build any software, because it makes few assumptions about what you're trying to build or how it should be done. The most notable restriction is that dependency management currently only supports Maven- and Ivy-compatible repositories and the filesystem.
+
+This doesn't mean you have to do a lot of work to create a build. Gradle makes it easy to build common types of project — say Java libraries — by adding a layer of conventions and prebuilt functionality through <<plugins#plugins,_plugins_>>. You can even create and publish custom plugins to encapsulate your own conventions and build functionality.
```
:x: "types of project_s_"

eriwen replied:
Would you say "projects" would be considered incorrect in British English? I do think this is considered fairly incorrect for American-style English.

pledbrook replied:
Yes, it sounds very odd. It seems to be a cultural thing, as witnessed by some of the comments on the [English StackExchange](https://english.stackexchange.com/questions/5539/types-of-things-vs-types-of-thing). I'll try to remember this for future reference, even if it hurts my internal ear :smile:

eriwen commented on subprojects/docs/src/docs/userguide/getting_started.adoc:
```diff
@@ -0,0 +1,69 @@
+ * _Runs on the JVM_ — but it can build anything, including native libraries and executables
+ * _Runs fast_ — particularly with well-designed builds
+ * _Uses conventions and plugins_ — which make it simple to build common types of project
+ * _Is customizable and extensible_
```
:o: Feels a little off that this is the only point without extra explanation.

pledbrook replied:
This is now in the _What is Gradle?_ chapter/page.

eriwen commented on subprojects/docs/src/docs/userguide/getting_started.adoc:
```diff
@@ -0,0 +1,69 @@
+
+ * <<five_things#five_things,5 things you need to know about Gradle>>
+
+== Short tutorials
```
It feels like we ought to have a brief "Installing" section similar to this: https://docs.bazel.build/versions/master/getting-started.html#installation

pledbrook replied:
Why have that in addition to the high-level _Installing Gradle_ page? Which is just below the _Getting Started_ one. I don't feel that strongly about it, but it seems a little overkill.

I can understand in one sense, because installation is typically part of getting started. OK, I'll probably include it.

Anyway, I think I need to give some thought and attention to this page. Most of my energy was focused on the Five Things and Groovy primer pages, although I should probably take another look at the latter too.

eriwen replied:
Great. Thanks for working through this. Overall we will produce something far superior to what we have today.

pledbrook commented on subprojects/docs/src/docs/userguide/getting_started.adoc:
```diff
@@ -0,0 +1,69 @@
+
+As with many flexible and customizable tools, Gradle can take a little time to understand. However, you can get a jump start by reading this short primer:
+
+ * <<five_things#five_things,5 things you need to know about Gradle>>
```
I wanted a call out similar to a quotation without the quotes. I just couldn't find anything I'm happy with. I'm not happy with the bullet point either! Not sure what to do here, but I'll review it again.

In general, I want a way to present standalone statements or principles that stand out without the decoration associated with notes and tips.

pledbrook replied:
This is now handled differently.

pledbrook commented:
@eriwen I couldn't reply directly to your comment where you say updating the home page is a blocker for 5.0. I think we'd already agreed to make changes to it, but you wanted it as a separate pull request if I remember correctly. Is that still the case? Or would you like to roll that change into this pull request since they're so tightly linked? (Note that I haven't spent time on the docs home page yet)

pledbrook commented:
@eriwen I made some substantial changes to the Getting Started page and added a "What is Gradle?" page that fleshes out the bullet points that were originally in the Getting Started page and user manual home. I've also moved the "5 Things" content from its own page to the "What is Gradle?" page because it felt like a natural fit.

I'm considering the addition of a "Why Gradle?" section to the "What is Gradle?" page, but I haven't convinced myself that it's a great idea yet.

Let me know what you think of the changes.

eriwen commented:
@pledbrook I think the getting started page is much better. Yes, let's ignore the navigation for now. It ties in with other PRs so let's not complicate this one. 

I can see how "5 Things" fits with "Getting Started". I'm torn what to do here because I feel a first-class "Key Gradle Concepts" page is something we've always been missing (words for thinking maybe: If you were presenting about Gradle to new developers, what bullet points would you use?)

I can definitely see "What Gradle does" and "Why use Gradle" being separate. When explaining Gradle to new folks at conferences, I have found it very helpful to start with "What" and then "Why". I'm for a separate subsection.

One subsection I think we might consider is "How to inspect your build" and/or something about troubleshooting. This fits under both running and authoring Gradle builds, so perhaps it deserves it's own section or ... I'm not sure how exactly it fits in but I think it's reasonably important (though not like a super blocker).

On the whole, the copy within each subsection is fairly good. I've read through it but want to ensure the organization is right and that I can read through it again before providing feedback on turns of phrase and punctation and the like.

pledbrook commented:
I would like to keep docs.gradle.org distinct from gradle.org. I view an approach along the lines of a presentation more appropriate for the latter, while I think the docs should focus on more in-depth information, with perhaps a slightly more neutral tone. That doesn't preclude covering the same topics and even material.

On the "5 Things" vs "Key Gradle Concepts", I favour the former because it's more forceful in asserting that the reader really should read it. Plus putting a small number on it makes it seem low effort to read. That said, perhaps there is scope for a more comprehensive Key Gradle Concepts page. I'm not even wedded to keeping "5 Things" with "What is Gradle?" if you'd prefer to have it as a separate page.

I'm all for something about troubleshooting builds, but I don't know what would go into it. And how would it relate to the existing Troubleshooting chapter (which could probably do with fleshing out itself)?

eriwen commented:
I have merged these changes to the release branch with https://github.com/gradle/gradle/commit/08ff46545680c9193916d1f759a5222ab6c4d45b
